src/audio_pixel.cpp

```diff
@@ -0,0 +1,183 @@
+        int collected_samples = 0;
+
+        for (int i = 0; i < sample_buffer.size(); i += num_channels) { 
+            double curr_sample = sample_buffer[i];
+
+            // increment collected samples after creating new audio pixel 
+            collected_samples ++;
+    
+            // update all fields of the current pixeld based on the current sample
+            curr_pixel.m_max = std::max(curr_pixel.m_max, curr_sample);
+            curr_pixel.m_min = std::min(curr_pixel.m_min, curr_sample);
+            curr_pixel.m_rms += (curr_sample * curr_sample);
+
+            if ((collected_samples % samples_per_pixel) == 0) {
+                // complete the RMS calculation, this method accounts of edge cases (total sample % samples per pixel != 0)
+                curr_pixel.m_rms = curr_pixel.m_rms / collected_samples;
+
+                // add the curr pixel to its channel in m_channel_pixels, clear curr_pixel
+                m_channel_pixels[channel].push_back(curr_pixel);
+
+                // set max to equal reaper min, and min to equal reaper max
+                curr_pixel = audio_pixel_t(DBL_MIN, DBL_MAX, 0);
+            }
+        }
+    }
+}
+
+// ************* audio_pixel_mipmap_t ****************
+
+audio_pixel_block_t audio_pixel_mipmap_t::get_pixels(double t0, double t1, double pix_per_s) {
+    // start by grabbing pixels from the nearest two pixel blocks TODO: this will be a vector<double>
+    double nearest_pps = get_nearest_pps(pix_per_s);
+
+    if (nearest_pps == pix_per_s) // last edge case, when hte given pps is already a block
+        return m_blocks[nearest_pps];
+
+    // perform interpolation
+    audio_pixel_block_t interpolated_block = interpolate_block(t0, t1, pix_per_s, nearest_pps);
+    return interpolated_block;
+};
+
+double audio_pixel_mipmap_t::get_nearest_pps(double pix_per_s) {
+    // TODO: make this a vector of doubles, return the two nearest pps
+    // returning the nearest to first get bilinear interpolation working
+    // 2 edge cases
+    int n_blocks = m_block_pps.size();
+
+    if (pix_per_s <= m_block_pps[0]) // given pps smaller than our smallest rep
+        return m_block_pps[0];
+    else if (pix_per_s >= m_block_pps[n_blocks - 1]) // given pps larger than our largest rep
+        return m_block_pps[n_blocks - 1];
+
+    return get_nearest_block_pps(pix_per_s);
+}
+
+double audio_pixel_mipmap_t::linear_interp(double x, double x1, double x2, double y1, double y2) {
+    // interpolation in time domain
+    return (((x2 - x) / (x2 - x1)) * y1) + ((x - x1) / (x2 - x1)) * y2;
+}
+
+audio_pixel_block_t audio_pixel_mipmap_t::interpolate_block(double t0, double t1, double new_pps, double nearest_pps) {
+    // interpolate from the block with nearest-pps to create a new block at new_pps
+    // the tricky part here is keeping track of the which samples from the nearest block are towards us
+    audio_pixel_block_t nearest_block = m_block_pps[nearest_pps];
```

Comment:
this pps does not necessarily need to be nearest_pps, at list within the scope of this function, whose purpose is to create a new block from a given source block, via linear interpolation. 

so we don't really need any of these `nearest` prefixes on the variable names (unless the variable actually relates to how near each individual sample is to each other within this linear interpolation logic).

Reply:
so i suggest a rename:

```cpp
audio_pixel_block_t interpolate_block(const audio_pixel_block_t& src, double new_pps, double t0, double t1)
```

Reply:
gotcha, thanks for this I'll try to think about the abstraction when writing a function

Reply:
for now I'm keeping the double value for source pps (now called `src_pps` instead of nearest), this is because it make's checking bounds easier if you have the src pps